pkg/clustermanager/controller.go

```diff
@@ -0,0 +1,188 @@
+	context context.Context,
+	clientset edgeclient.Interface,
+	k8sClientset *kubeclient.Clientset,
+	logicalClusterInformer cache.SharedIndexInformer,
+	providerInformer cache.SharedIndexInformer,
+) *controller {
+	context = klog.NewContext(context, klog.FromContext(context).WithValues("controller", controllerName))
+
+	c := &controller{
+		context:                 context,
+		clientset:               clientset,
+		k8sClientset:            k8sClientset,
+		logger:                  klog.FromContext(context),
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		logicalClusterInformer:  logicalClusterInformer,
+		clusterProviderInformer: providerInformer,
+		providers:               make(map[string]*provider),
+	}
+
+	logicalClusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueLogicalCluster,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			old := oldObj.(*lcv1alpha1.LogicalCluster)
+			new := newObj.(*lcv1alpha1.LogicalCluster)
+			if !reflect.DeepEqual(old, new) {
+				c.enqueueLogicalCluster(newObj)
+			}
+		},
+		DeleteFunc: c.enqueueLogicalCluster,
+	})
+
+	providerInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueueClusterProviderDesc,
+		DeleteFunc: c.enqueueClusterProviderDesc,
+	})
+
+	return c
+}
+
+func (c *controller) enqueueLogicalCluster(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	c.logger.V(2).Info("queueing LogicalCluster", "key", key)
+	c.queue.Add(
+		queueItem{
+			triggeringKind: triggeringKindLogicalCluster,
+			key:            key,
+		},
+	)
+}
+
+func (c *controller) enqueueClusterProviderDesc(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	c.logger.V(2).Info("queueing ClusterProviderDesc", "key", key)
+	c.queue.Add(
+		queueItem{
+			triggeringKind: triggeringKindClusterProviderDesc,
+			key:            key,
+		},
+	)
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Info("starting manager logicalcluster controller")
+	defer c.logger.Info("shutting down manager logicalcluster controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.context, c.runWorker, time.Second)
+	}
+
+	<-c.context.Done()
+}
+
+func (c *controller) runWorker(ctx context.Context) {
+	for c.processNextItem() {
+	}
+}
+
+func (c *controller) processNextItem() bool {
+	// Wait until there is a new item in the working queue
+	i, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	item := i.(queueItem)
+
+	// Done with this key, unblock other workers.
+	defer c.queue.Done(i)
+
+	if err := c.process(item); err != nil {
+		runtime.HandleError(err)
```

Comment:
`runtime.HandleError` is for un-handled errors.  I see this error being handled in the next line.

Reply:
Handled by PR #816